File: designParkingLot.py
"""
Time Complexity : 
1. addParkingSlot - O(nlog n) where n is the total number of parking slots added nlogn is the time complexity for the heapify which will take place at the time when we will add any element to the heap.
2.getAvailability : O(1)
3. park : O(1), need to remove the parking slot at the top of the heap.
4. unPark: O(nlog n) where n is the total number of parking slots added nlogn is the time complexity for the heapify which will take place at the time when we will add any element to the heap.
                  
Space Complexity : O(n) which is the number of elements present in the heap. In worst case m * n where m is the maximum number of floors and n is the maximum number of spots per floor.

Did this code successfully run on Leetcode : Yes
Any problem you faced while coding this : No

Your code here along with comments explaining your approach:
As this is the design problem, start thinking from the bigger view
Person enters the parking Lot, looks for the nearest possible empty spot, once get the nearest possible empty spot parks the vehicle and get the token. After some time leaves the parking slot, so that spot will be available for the parking, compare it with the other avaialble parking slots and keep ready the nearest empty spot for the next person.
"""
import heapq

class ParkingSpot:
    def __init__(self, floor, spot):
        self.floor = floor
        self.spot = spot
    
    def __lt__(self, other):
        if self.floor == other.floor:
            return self.spot < other.spot
        return self.floor < other.floor
    
    def getFloor(self):
        return self.floor
    
    def getSpot(self):
        return self.spot
    
class ParkingLot:
    def __init__(self, maxFloor, spotsPerFloor):
        self.maxFloors = maxFloor
        self.spotsPerFloor = spotsPerFloor
        self.heap = []
    
    def addParkingSlot(self, floor, spot):
        if floor > self.maxFloors:
            raise Exception("Floor number is greater than the maximum available floors")
        if spot > self.spotsPerFloor:
            raise Exception("Spot number is greater than the maximum available spots per floor")
        newParkingSpot = ParkingSpot(floor, spot)
        heapq.heappush(self.heap, newParkingSpot)
    
    def getAvailability(self):
        if len(self.heap) == 0:
            raise Exception("No Parking Spots are avaialble")
        availableParkingSlot = self.heap[0]
        return availableParkingSlot
    
    def park(self):
        return heapq.heappop(self.heap)
    
    def unPark(self, floor, spot):
        newParkingSpot = ParkingSpot(floor, spot)
        heapq.heappush(self.heap, newParkingSpot)

def main():

    pl = ParkingLot(10, 20)
    pl.addParkingSlot(1, 1)
    pl.addParkingSlot(2, 1)
    pl.addParkingSlot(3, 1)
    pl.addParkingSlot(1, 2)
    pl.addParkingSlot(2, 2)
    pl.addParkingSlot(3, 2)
    
    nextAvailableParkingSpot = pl.getAvailability()
    print(f"Floor: {nextAvailableParkingSpot.floor}")
    print(f"Spot: {nextAvailableParkingSpot.spot}")
    
    pl.park()
    
    nextAvailableParkingSpot = pl.getAvailability()
    print(f"Floor: {nextAvailableParkingSpot.floor}")
    print(f"Spot: {nextAvailableParkingSpot.spot}")
    pl.unPark(1, 1)
    
    nextAvailableParkingSpot = pl.getAvailability()
    print(f"Floor: {nextAvailableParkingSpot.floor}")
    print(f"Spot: {nextAvailableParkingSpot.spot}")

main()
    
    
    

File: DesignParkingLot.cs
using System;
using System.Collections.Generic;
using System.Text;

namespace Design
{
    class ParkingLotLC
    {
        //TC: Except for GetAvailableSlots, it is O(logN). For GetAvailableSlots it is O(1)
        //SC: Priority Queue size
        //.NET do not have PriorityQueue in place
        public class ParkingLot
        {
            public int floors, spots;
            PriorityQueue<ParkingSpot> pq;
            public ParkingLot(int floors, int spots)
            {
                this.floors = floors;
                this.spots = spots;
                //pq = new PriorityQueue<ParkingSpot>((a, b) =>
                //                                    {
                //                                        if (a.getFloor() == b.getFloor())
                //                                        {
                //                                            return a.getSpot() - b.getSpot();
                //                                        }
                //                                        return a.getFloor() - b.getFloor();
                //                                    });
            }

            //parking the vehicle
            public ParkingSpot park()
            {
                getAvailableParkingSpot();
                ParkingSpot p = pq.Dequeue();
                return p;
            }

            //getting available parking spots
            public ParkingSpot getAvailableParkingSpot()
            {
                if (pq.Count == 0)
                {
                    throw new Exception("No parking spots are available");
                }
                return pq.Peek();
            }
            //unpark the vehicle
            public void unPark(int floor, int spot)
            {
                if (floor > floors)
                {
                    throw new Exception("Invalid floor");
                }
                if (spot > spots)
                {
                    throw new Exception("Invalid spot");
                }
                pq.Enqueue(new ParkingSpot(floor, spot));
            }
            //add parking space
            public void addParkingSpace(int floor, int spot)
            {
                if (floor > floors)
                {
                    throw new Exception("Invalid floor");
                }
                if (spot > spots)
                {
                    throw new Exception("Invalid spot");
                }
            }
        }

        public  class ParkingSpot
        {
            internal int floor, spot;
            public ParkingSpot(int floor, int spot)
            {
                this.floor = floor;
                this.spot = spot;
            }
            public int getFloor()
            {
                return this.floor;
            }
            public int getSpot()
            {
                return this.spot;
            }
        }

    }
}
